segmentstore/server/src/main/java/io/pravega/service/server/logs/DurableLog.java

```diff
@@ -234,8 +235,11 @@ public int getId() {
         TimeoutTimer timer = new TimeoutTimer(timeout);
         log.info("{}: Truncate (OperationSequenceNumber = {}, DataFrameAddress = {}).", this.traceObjectId, upToSequenceNumber, truncationFrameAddress);
 
-        return this.durableDataLog
-                .truncate(truncationFrameAddress, timer.getRemaining())
+        // Before we do any real truncation, we need to mini-snapshot the metadata with only those fields that are updated
+        // asynchronously for us (i.e., not via normal Log Operations) such as the Storage State. That ensures that this
+        // info will be readily available upon recovery without delay.
+        return add(new StorageMetadataCheckpointOperation(), timer.getRemaining())
```

Comment:
I have a few clarification questions here:
1- We already have `MetadataCheckpointOperation` and now we are adding `StorageMetadataCheckpoint` operation. As I understand it, the former is Tier 1 metadata while the latter is Tier 2 metadata?
2- Could we consolidate `MetadataCheckpointOperation` and `StorageMetadataCheckpoint`?
3- The comment block says: _we need to mini-snapshot the metadata with only those fields that are updated asynchronously for us_. Because it is asynchronous, is it possible that the storage metadata will contain stale data?
4- I don't actually get whether we are guaranteeing that the storage checkpoint is being flushed before the truncation. If we truncate without flushing, then it is possible that a crash will cause us to lose that checkpoint.